contrib/p4/p4execution.py

```diff
@@ -0,0 +1,838 @@
+# SPDX-FileCopyrightText: 2020-2021 CASTOR Software Research Centre
+# <https://www.castor.kth.se/>
+# SPDX-FileCopyrightText: 2020-2021 Johan Paulsson
+
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+import os
+import logging
+
+import time
+import re
+import threading
+from benchexec import systeminfo
+from benchexec.model import Run
+from p4.p4_run_setup import P4SetupHandler
+from p4.Counter import Counter
+
+from benchexec import tooladapter
+from benchexec import util
+from benchexec import benchexec
+from benchexec import BenchExecException
+
+# File handling
+import zipfile
+from shutil import copyfile, rmtree
+import json
+from distutils.dir_util import copy_tree
+
+try:
+    import docker
+except Exception as e:
+    raise BenchExecException(
+        "Python-docker package not found. Try reinstalling python docker module"
+    )
+
+try:
+    from pyroute2 import IPRoute
+    from pyroute2 import NetNS
+except:
+    raise BenchExecException(
+        "pyroute2 python package not found. Try reinstalling pyroute2"
+    )
+
+
+STOPPED_BY_INTERRUPT = False
+
+# Static Parameters
+MGNT_NETWORK_SUBNET = "172.19"  # Subnet 192.19.x.x/16
+NODE_IMAGE_NAME = "basic_node"
+SWITCH_IMAGE_NAME = "switch_bmv2"
+PTF_IMAGE_NAME = "ptf_tester"
+
+
+class P4Execution(object):
+    """
+    This Class is for executing p4 benchmarks. The class creates docker containers representing each
+    device in the network. It creates virutal ethenet connections between all the devices. Finally,
+    it sets up a test container connected to all the nodes in the network.
+    """
+
+    def __init__(self):
+        self.nodes = None  # Set by init
+        self.switches = None  # Set by init
+        self.ptf_tester = None  # Set by init
+
+        # Includes all nodes and switches, not the ptf tester
+        self.nr_of_active_containers = Counter()
+
+        self.client = None
+        self.node_networks = []
+        self.mgnt_network = None
+
+    def init(self, config, benchmark):
+        """
+        This functions will set up the docker network to execute the test.
+        As a result, it needs root permission for the setup part.
+        """
+
+        tool_locator = tooladapter.create_tool_locator(config)
+        benchmark.executable = benchmark.tool.executable(tool_locator)
+        benchmark.tool_version = benchmark.tool.version(benchmark.executable)
+
+        # Read test inputs paths
+        (
+            self.switch_source_path,
+            self.ptf_folder_path,
+            self.network_config_path,
+        ) = self.read_folder_paths(benchmark)
+
+        if not os.path.isdir(self.switch_source_path):
+            logging.critical(f"Switch folder path not found: {self.switch_source_path}")
```

Comment:
Hint: for logging, the old `%s` style formatting is often nicer because it is built in into the `logging` module:
```suggestion
            logging.critical("Switch folder path not found: %s", self.switch_source_path)
```
In all other cases, f-strings are of course nicer.